feat(window-menu): add docs for the `menu` (#2875)
tauri-apps/tauri-docs: src/content/docs/plugin/window-menu.mdx

---

> title: Window menu

Reviewer: ```suggestion
title: Window Menu
```

tableOfContents:
  maxHeadingLevel: 4
---

import { Tabs, TabItem } from '@astrojs/starlight/components';

> Native application menus can be attached to a window or can be attached to the system tray too. Available on desktop only.

Reviewer: ```suggestion
Native application menus can be attached to both to a window or system tray. Available on desktop.
```

> ## Creating a base-level menu
> 
> To create a base-level native window menu, and attached to the window.

Reviewer: ```suggestion
## Creating a base-level menu

To create a base-level native window menu, and attach to a window:
```
It could be "**a** window" or "**all** windows". Is there any difference?

Author: There seems to be no difference in the adjectives, but I remember there is an API that can add different effects to a `window menu` for a specific window.

<Tabs>
<TabItem label="JavaScript">

Use the [`Menu.new`] static function to create a window menu:

```javascript
import { Menu } from '@tauri-apps/api/menu';

const menu = await Menu.new({
  items: [
    {
      id: 'quit',
      text: 'Quit',
      action: () => {
        console.log('quit pressed');
      },
    },
  ],
});

> //  If a window was not created with an explicit menu or had one set explicitly,

Reviewer: ```suggestion
// If a window was not created with an explicit menu or had one set explicitly,
```
Same, is this window specific? What if I want to attach to all windows? Or is it always to all of them?

// this menu will be assigned to it.
menu.setAsAppMenu().then((res) => {
  console.log('menu set success', res);
});
```

</TabItem>

<TabItem label="Rust">

```rust
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
use tauri::menu::{Menu, MenuItem};

fn main() {
  tauri::Builder::default()
        .setup(|app| {
            let menu = MenuBuilder::new(app)
                .text("open", "Open")
                .text("close", "Close")
                .build()?;

            app.set_menu(menu)?;

            Ok(())
        })
}
```

</TabItem>
</Tabs>

## Listening to events on custom menu items

> Each custom menu item triggers an event when clicked. Use the on_menu_event API to handle them.

Reviewer: ```suggestion
Each custom menu item triggers an event when clicked. Use the `on_menu_event` API to handle them.
```
do you know if is there any doc reference to `on_menu_event`? Also since it's Rust specific, would be better to move it to be visible only when "Rust" tab is active

      },
    },
    {
      id: 'Close',
      text: 'close',
      action: () => {
        console.log('close pressed');
      },
    },
  ],
});

await menu.setAsAppMenu();
```

</TabItem>

<TabItem label="Rust">

```rust
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
use tauri::menu::{MenuBuilder};

fn main() {
  tauri::Builder::default()
        .setup(|app| {
            let menu = MenuBuilder::new(app)
                .text("open", "Open")
                .text("close", "Close")
                .build()?;

            app.set_menu(menu)?;

            app.on_menu_event(move |app_handle: &tauri::AppHandle, event| {

                println!("menu event: {:?}", event.id());

                match event.id().0.as_str() {
                    "open" => {
                        println!("open event");
                    }
                    "close" => {
                        println!("close event");
                    }
                    _ => {
                        println!("unexpected menu event");
                    }
                }
            });

            Ok(())
        })
}
```

</TabItem>
</Tabs>

## Creating a multi-level menu

> To create a multi-level menu, you can add some submenus to the menu item.

Reviewer: ```suggestion
To create a multi-level menu, you can add some submenus to the menu item:
```

<Tabs>
<TabItem label="JavaScript">

```javascript
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use tauri::{image::Image, menu::{CheckMenuItemBuilder, IconMenuItemBuilder, MenuBuilder, SubmenuBuilder}};

> fn main() {

Reviewer: in this section both tabs has Rust code in it

                .text("open", "Open")
                .text("quit", "Quit")
                .build()?;

            let lang_str = "en";
            let check_sub_item_1 = CheckMenuItemBuilder::new("English")
                .id("en")
                .checked(lang_str == "en")
                .build(app)?;

            let check_sub_item_2 = CheckMenuItemBuilder::new("Chinese")
                .id("en")
                .checked(lang_str == "en")
                .enabled(false)
                .build(app)?;

             // Load icon from path
            let icon_image = Image::from_bytes(include_bytes!("../icons/icon.png")).unwrap();

            let icon_item = IconMenuItemBuilder::new("icon")
                .icon(icon_image)
                .build(app)?;

            let other_item = SubmenuBuilder::new(app, "language")
                .item(&check_sub_item_1)
                .item(&check_sub_item_2)
                .build()?;

            let menu = MenuBuilder::new(app)
                .items(&[&file_menu, &other_item,&icon_item])
                .build()?;

            app.set_menu(menu)?;

            Ok(())
        })
}
```

> Note that you need the image-ico or image-png Cargo features to use this API. To enable it, change your Cargo.toml file:

Reviewer: ```suggestion
Note that you need to enable `image-ico` or `image-png` feature to use this API:
```

> `src-tauri/Cargo.toml`
> 
> ```toml
> [dependencies]
> tauri = { version = "...", features = ["...", "image-png"] }
> ```

Reviewer: ```suggestion
```toml title="src-tauri/Cargo.toml"
[dependencies]
tauri = { version = "...", features = ["...", "image-png"] }
```
```

</TabItem>
</Tabs>

## Creating predefined menu

> To create predefined (native) menu item which has a predefined behavior by the OS or by tauri.

Reviewer: ```suggestion
To use built-in (native) menu items that has predefined behavior by the operating system or Tauri:
```

});

const paste = await PredefinedMenuItem.new({
  text: 'paste-text',
  item: 'Paste',
});

const select_all = await PredefinedMenuItem.new({
  text: 'select_all-text',
  item: 'SelectAll',
});

const menu = await Menu.new({
  items: [copy, separator, undo, redo, cut, paste, select_all],
});

await menu.setAsAppMenu();
```

</TabItem>

<TabItem label="Rust">

```rust
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
use tauri::menu::{MenuBuilder, PredefinedMenuItem};

fn main() {
  tauri::Builder::default()
        .setup(|app| {
      let menu = MenuBuilder::new(app)
                .copy()
                .separator()
                .undo()
                .redo()
                .cut()
                .paste()
                .select_all()
                .item(&PredefinedMenuItem::copy(app, Some("custom text"))?)
                .build()?;
            app.set_menu(menu)?;

            Ok(())
        })
}
```

For more preset capabilities, please refer to the documentation [`PredefinedMenuItem`].

:::tip
The menu builder has dedicated methods to add each predefined menu item so you can call `.copy()` instead of `.item(&PredefinedMenuItem::copy(app, None)?)`.

> :::

Reviewer: If possible move to the Rust tab, I'm not sure if it would look good

</TabItem>
</Tabs>

## Change menu status

> If you want to change the status of the menu, such as text, icon, check status, you should `set_menu` again.

Reviewer: ```suggestion
If you want to change the status of the menu, such as text, icon, or check status, you can `set_menu` again:
```
